### PR TITLE
UN-2239 [FEAT] Add trial statistics API endpoint for organization usage tracking

### PR DESCRIPTION
**Pull request by jaags-dev**
## What
Added a new API endpoint `/api/v2/usage/trial-statistics/` that retrieves comprehensive trial usage statistics for organizations. This endpoint provides key metrics about an organization's trial period usage including costs, document processing counts, API calls, and ETL pipeline runs.

## Why
This endpoint is needed to provide organizations with insights into their trial period usage patterns. It enables tracking of resource consumption, cost savings, and usage metrics during the trial phase, which is crucial for understanding customer engagement and conversion potential.

## How
- **UsageHelper Enhancement**: Added `get_trial_statistics()` static method that:
  - Attempts to fetch subscription data from the subscription plugin if available
  - Falls back to organization creation date + 14 days for trial period calculation
  - Aggregates usage data using Django ORM for efficient queries
  - Calculates total cost, document processing counts, API calls, and unique ETL runs
  - Includes comprehensive error handling with fallback responses

- **View Implementation**: Added `get_trial_statistics` action in UsageView that:
  - Retrieves the current organization context
  - Calls the helper method to get statistics
  - Returns properly formatted JSON response with error handling

- **URL Configuration**: Registered new `/trial-statistics/` endpoint in the usage_v2 URL patterns

## Relevant Docs
- Endpoint: `GET /api/v2/usage/trial-statistics/`
- Response format:
  ```json
  {
    "trial_start_date": "ISO date string",
    "trial_end_date": "ISO date string",
    "total_cost": 0.0,
    "documents_processed": 0,
    "api_calls": 0,
    "etl_runs": 0
  }
  ```

## Related Issues/PRs
- Related to UN-2239 ticket for trial statistics tracking
- Builds upon existing usage_v2 infrastructure

## Dependencies Versions
No new dependencies added. Uses existing Django ORM and subscription plugin (optional).

### Manual Testing Scenarios:
1. **With Subscription Plugin:**
   - Verify endpoint returns subscription-based trial dates
   - Confirm accurate cost and usage aggregations

2. **Without Subscription Plugin (Fallback):**
   - Verify fallback to organization creation date + 14 days
   - Ensure no errors when subscription plugin is unavailable

3. **Edge Cases:**
   - Test with organizations having no usage data
   - Test with organizations having partial data
   - Verify error handling returns appropriate fallback response

### Test Commands:
```bash
# Test the endpoint
curl -X GET http://localhost:8000/api/v2/usage/trial-statistics/ \
  -H "Authorization: Bearer <token>"
```

### Validation Checklist:
- [ ] Endpoint returns 200 status with valid data
- [ ] Dates are in ISO format
- [ ] Numeric values are properly calculated
- [ ] Error scenarios return graceful fallback data
- [ ] Authentication and organization context work correctly

## Screenshots
N/A - Backend API endpoint only, no UI changes

### CODE REVIEW

File: backend/usage_v2/views.py [hunk 1]
```diff
@@ -93,3 +93,31 @@ def get_token_usage(self, request: HttpRequest) -> Response:
...
+        try:
+            user_organization = UserContext.get_organization()
+
+            # Get trial statistics from helper
+            trial_stats = UsageHelper.get_trial_statistics(user_organization)
+
+            return Response(status=status.HTTP_200_OK, data=trial_stats)
+
+        except Exception as e:
+            logger.error(f"Error retrieving trial statistics: {str(e)}")
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/views.py]: @jagadeeswaran-zipstack avoid such catch all exception handling. This will just suppress the error and not allow us to print a valid stack trace. Remove this and check what happens in case of an error

**Review by jaseemjaskp — APPROVED** [review]

**Review by Deepak-Kesavan — APPROVED** [review]

File: backend/usage_v2/views.py [hunk 2]
```diff
@@ -93,3 +93,60 @@ def get_token_usage(self, request: HttpRequest) -> Response:
...
+            return Response(
+                status=status.HTTP_400_BAD_REQUEST,
+                data={"error": "Invalid organization data"},
+            )
+        except Exception as e:
+            logger.error(f"Unexpected error retrieving trial statistics: {str(e)}")
+            return Response(
+                status=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                data={"error": "Unable to retrieve trial statistics"},
+            )
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/views.py]:
```suggestion
```
@jagadeeswaran-zipstack this could be problematic. If you raise a valid exception, it will be caught here and a different message is going to be raised. I suggest removing this altogether

File: backend/usage_v2/views.py [hunk 3]
```diff
@@ -93,3 +93,60 @@ def get_token_usage(self, request: HttpRequest) -> Response:
...
+
+            return Response(status=status.HTTP_200_OK, data=trial_stats)
+
+        except ValueError as e:
+            logger.error(f"Validation error retrieving trial statistics: {str(e)}")
+            return Response(
+                status=status.HTTP_400_BAD_REQUEST,
+                data={"error": "Invalid organization data"},
+            )
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/views.py]:
```suggestion
```
Do not frame your own error messages - we are using a library that converts all such python exceptions and handling the parsing in the FE accordingly. In order to stay consistent, raise this as a DRF exception with the correct status code and message. Such 4xx errors can usually be checked and raised at the start of API processing / serializer itself

File: backend/usage_v2/views.py [hunk 4]
```diff
@@ -93,3 +93,60 @@ def get_token_usage(self, request: HttpRequest) -> Response:
...
+        try:
+            user_organization = UserContext.get_organization()
+
+            # Validate organization context
+            if not user_organization:
+                logger.warning("No organization context found for user")
+                return Response(
+                    status=status.HTTP_400_BAD_REQUEST,
+                    data={"error": "No organization context available"},
+                )
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/views.py]:
```suggestion
```
Raise it as a valid DRF error instead to be consistent

File: backend/usage_v2/views.py [hunk 5]
```diff
@@ -93,3 +93,60 @@ def get_token_usage(self, request: HttpRequest) -> Response:
...
+
+            # If helper returns an error payload, surface it as 500 to clients.
+            if isinstance(trial_stats, dict) and trial_stats.get("error"):
+                logger.error(
+                    "Trial statistics helper returned error payload; responding with 500"
+                )
+                return Response(
+                    status=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                    data={"error": trial_stats["error"]},
+                )
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/views.py]:
```suggestion
```
Correct this - helper should raise a DRF exception itself

File: backend/usage_v2/helper.py [hunk 6]
```diff
@@ -98,3 +143,177 @@ def format_usage_response(
...
+                - trial_end_date (str): ISO formatted trial end date
+                - total_cost (float): Total cost in dollars during trial period
+                - documents_processed (int): Number of unique document processing operations
+                - api_calls (int): Total number of API calls made
+                - etl_runs (int): Number of unique ETL pipeline runs
+                - error (str, optional): Error message if processing fails
+        """
+        try:
+            # Validate organization
+            UsageHelper._validate_organization(organization)
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/helper.py]: @jagadeeswaran-zipstack I doubt if this is needed since we get the org_id from the DB which is already validated when stored. Also keep in mind to make such validations in a serializer

File: backend/usage_v2/helper.py [hunk 7]
```diff
@@ -12,6 +13,50 @@
 
 
 class UsageHelper:
+    @staticmethod
+    def normalize_to_timezone_aware_datetime(value, is_end_date=False, reference_tz=None):
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/helper.py]: @jagadeeswaran-zipstack please check if there's a library or if django does this for you instead of us attempting

File: backend/usage_v2/helper.py [hunk 8]
```diff
@@ -98,3 +143,177 @@ def format_usage_response(
...
+                - api_calls (int): Total number of API calls made
+                - etl_runs (int): Number of unique ETL pipeline runs
+                - error (str, optional): Error message if processing fails
+        """
+        try:
+            # Validate organization
+            UsageHelper._validate_organization(organization)
+
+            # Get reference timezone
+            reference_tz, org_created_at = UsageHelper._get_reference_timezone(
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/helper.py]: @jagadeeswaran-zipstack why do we do this? Aren't we storing the data uniformly in this table?

File: backend/usage_v2/views.py (hunk 2 above)
Comment (jaags-dev) [reply backend/usage_v2/views.py]: Understood, Thanks. I'll make the required change.

File: backend/usage_v2/helper.py (hunk 6 above)
Comment (jaags-dev) [reply backend/usage_v2/helper.py]: Removed the entire _validate_organization() method and its call. Yes, the organization is already validated when stored in DB.

File: backend/usage_v2/helper.py (hunk 7 above)
Comment (jaags-dev) [reply backend/usage_v2/helper.py]: the code already uses Django's built-in timezone.make_aware() and timezone.localtime() utilities. Added comment explaining the wrapper is needed for date-to-datetime conversion and end-of-day handling.

File: backend/usage_v2/helper.py (hunk 8 above)
Comment (jaags-dev) [reply backend/usage_v2/helper.py]: Yes, We dont need this check here.

File: backend/usage_v2/helper.py [hunk 9]
```diff
@@ -145,63 +101,49 @@ def format_usage_response(
...
     @staticmethod
-    def _validate_organization(organization):
-        """Validate organization has required attributes."""
-        if not organization:
-            raise ValueError("Organization cannot be None")
-        if not hasattr(organization, "organization_id"):
-            raise ValueError("Organization must have organization_id attribute")
-        if not hasattr(organization, "created_at"):
-            raise ValueError("Organization must have created_at attribute")
+    def _get_subscription_dates(organization_id):
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/helper.py]: NIT: Try to add typing information wherever possible

File: backend/usage_v2/helper.py [hunk 10]
```diff
@@ -274,32 +185,26 @@ def get_trial_statistics(organization) -> dict[str, Any]:
...
-                    reference_tz=reference_tz,
+                # For end date, set to end of day (23:59:59.999999)
+                end_of_trial_date = organization.created_at + timedelta(days=14)
+                trial_end_date = end_of_trial_date.replace(
+                    hour=23, minute=59, second=59, microsecond=999999
                 )
 
+            # Validate trial window - guard against inverted date range
+            if trial_end_date < trial_start_date:
+                raise ValueError("trial_end_date must be on or after trial_start_date")
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/helper.py]: @jagadeeswaran-zipstack this value error is a pythonic error which will not be reported correctly. You should raise a ValidationError and ensure that the status code is 400

File: backend/usage_v2/helper.py [hunk 11]
```diff
@@ -314,6 +219,6 @@ def get_trial_statistics(organization) -> dict[str, Any]:
                 "etl_runs": aggregated_data.get("unique_runs", 0) or 0,
             }
 
-        except Exception as e:
-            logger.error(f"Error calculating trial statistics: {str(e)}")
-            return UsageHelper._get_error_response(organization)
+        except Exception:
+            logger.exception("Error calculating trial statistics")
+            raise
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/helper.py]: @jagadeeswaran-zipstack if we're only catching to log and re-raise, you can simply remove this try-except.

File: backend/usage_v2/views.py [hunk 12]
```diff
@@ -113,40 +114,18 @@ def get_trial_statistics(self, request: HttpRequest) -> Response:
...
-            )
-        except Exception as e:
-            logger.error(f"Unexpected error retrieving trial statistics: {str(e)}")
-            return Response(
-                status=status.HTTP_500_INTERNAL_SERVER_ERROR,
-                data={"error": "Unable to retrieve trial statistics"},
-            )
+        except (ValueError, ValidationError):
+            logger.exception("Validation error retrieving trial statistics")
+            raise
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/views.py]: Remove this try-except if we are only going to log. In fact, we might see repeated logs / tracebacks. Try to induce an error and check how it gets reported without this

File: backend/usage_v2/helper.py [hunk 13]
```diff
@@ -145,63 +101,49 @@ def format_usage_response(
...
         try:
             from pluggable_apps.subscription_v2.subscription_helper import (
                 SubscriptionHelper,
             )
+        except ImportError:
+            logger.debug("Subscription plugin not available, using fallback dates")
+            return None, None
 
+        # Plugin is available, attempt to get subscription data
+        try:
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/helper.py:134]: NIT: Move this plugin import at the top and in case of an import error, mark this helper as None. Exit early from this function if its not present

**Review by chandrasekharan-zipstack — CHANGES_REQUESTED** [review]
Left some more comments, ideally this should have been pluggable

File: backend/usage_v2/helper.py (hunk 9 above)
Comment (jaags-dev) [reply backend/usage_v2/helper.py]: Done 👍

File: backend/usage_v2/helper.py (hunk 10 above)
Comment (jaags-dev) [reply backend/usage_v2/helper.py]: Understood. Made the changes.

File: backend/usage_v2/helper.py (hunk 13 above)
Comment (jaags-dev) [reply backend/usage_v2/helper.py:134]: Done 👍

File: backend/usage_v2/helper.py [hunk 14]
```diff
@@ -4,11 +4,21 @@
...
 
+# Try to import subscription plugin at module level
+try:
+    from pluggable_apps.subscription_v2.subscription_helper import (
+        SubscriptionHelper,
+    )
+
+    SUBSCRIPTION_HELPER = SubscriptionHelper
+except ImportError:
+    SUBSCRIPTION_HELPER = None
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/helper.py:20]:
```suggestion
except ImportError:
    SubscriptionHelper = None
```

File: backend/usage_v2/helper.py [hunk 15]
```diff
@@ -101,23 +111,28 @@ def format_usage_response(
...
+        Returns:
+            Tuple of (start_date, end_date) or (None, None) if not available
         """
-        try:
-            from pluggable_apps.subscription_v2.subscription_helper import (
-                SubscriptionHelper,
-            )
-        except ImportError:
+        # Exit early if subscription plugin is not available
+        if SUBSCRIPTION_HELPER is None:
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/helper.py:129]:
```suggestion
        if SubscriptionHelper is None:
```

File: backend/usage_v2/helper.py [hunk 16]
```diff
@@ -101,23 +111,28 @@ def format_usage_response(
...
             logger.debug("Subscription plugin not available, using fallback dates")
             return None, None
 
         # Plugin is available, attempt to get subscription data
         try:
-            org_plans = SubscriptionHelper.get_subscription(organization_id)
+            org_plans = SUBSCRIPTION_HELPER.get_subscription(organization_id)
```
Comment (chandrasekharan-zipstack) [inline backend/usage_v2/helper.py:135]:
```suggestion
            org_plans = SubscriptionHelper.get_subscription(organization_id)
```
